### DISCUSSION TITLE
Asserting against cookies before redirect occurs

### DISCUSSION THREAD

**Discussion by jchi2241** [cypress-io/cypress#17341] [General]
I have a form on a landing page that I would like to test. It executes these steps in the following order when the submit button is clicked:

1. posts an endpoint. when the response is received, it:
2. sets a cookie
3. redirects to a website on another domain

something like
```javascript
function handleSubmit () {
  postEndpoint(...).then((success) => {
    if (success) {
       setCookie("blah");
       navigate(...);
     }
  });
}
```

I am currently attempting to test that the cookie does get set before it redirects. Something like
```javascript
cy.intercept(...).as("endpoint");
cy.visit(...);
cy.clickSubmitButton();
cy.getCookie("blah").then(cookie => {
   expect(...)
});
cy.wait("@endpoint");
cy.assertPageRedirectsTo(...);
```
The problem I'm encountering is that when `cy.getCookie()` runs, it does not get the latest state of the cookie. However, if I place `cy.getCookie()` after `cy.wait("@endpoint")`, the page has already redirected, and Cypress is calling `cy.getCookie` on the redirected page on the other domain, where the cookie obviously does not exist.

I've also tried using the `window:before:unload` event and wrapping `cy.getCookie` to "wait" for the cookie to set by only checking for the cookie before the page unloads
```javascript
cy.on("window:before:unload", () => {
  cy.getCookie("blah").then(cookie => {
     expect(...)
  });
});
```
but Cypress complains with the following error:
```
Cypress detected that you returned a promise from a command while also invoking one or more cy commands in that promise.

The command that returned the promise was:

  > cy.visit()

The cy command you invoked inside the promise was:

  > cy.getCookie()
```

Does anyone have suggestions on how to tackle this problem? Does some sort of "before intercept" hook exist? Is my implementation unusual?

**Comment by herecydev** [comment]
I've stumbled on a similar problem, so I don't think it's an isolated issue. In my case, we have a landing page where a user might land on, and would quickly expect a redirect to happen. It's still important that we present some information to the user (loading spinner, progress bar, etc. ) and would like to test that.

The issue is similar; how do we use the cypress api before the redirect happens. There's a race between cypress finishing and the application redirecting. I've definitely tried using some of the cypress events (like you have) but it doesn't expect you to use `cy.` I believe, hence the error.
